src/registrar/models/domain.py

```diff
@@ -667,6 +649,200 @@ def __str__(self) -> str:
+    # This can likely be done without passing in
+    # contact_id and contact_type and instead embedding it inside of
+    # contact, but the tradeoff for that is that it unnecessarily complicates using this
+    # (as you'd have to create a custom dictionary), and type checking becomes weaker.
+    # I'm sure though that there is an easier alternative...
+    # TLDR: This doesn't look as pretty, but it makes using this function easier
+    def map_epp_contact_to_public_contact(
+        self,
+        contact: eppInfo.InfoContactResultData,
+        contact_id,
+        contact_type,
+        create_object=True,
+    ):
+        """Maps the Epp contact representation to a PublicContact object.
+
+        contact -> eppInfo.InfoContactResultData: The converted contact object
+
+        contact_id -> str: The given registry_id of the object (i.e "cheese@cia.gov")
+
+        contact_type -> str: The given contact type, (i.e. "tech" or "registrant")
+
+        create_object -> bool: Flag for if this object is saved or not
+        """
+
+        if contact is None:
+            return None
+
+        if contact_type is None:
+            raise ValueError("contact_type is None")
+
+        if contact_id is None:
+            raise ValueError("contact_id is None")
+
+        if len(contact_id) > 16 or len(contact_id) < 1:
+            raise ValueError(
+                "contact_id is of invalid length. "
+                "Cannot exceed 16 characters, "
+                f"got {contact_id} with a length of {len(contact_id)}"
+            )
+
+        if not isinstance(contact, eppInfo.InfoContactResultData):
+            raise ValueError("Contact must be of type InfoContactResultData")
+
+        auth_info = contact.auth_info
+        postal_info = contact.postal_info
+        addr = postal_info.addr
+        # 'zips' two lists together.
+        # For instance, (('street1', 'some_value_here'),
+        # ('street2', 'some_value_here'))
+        # Dict then converts this to a useable kwarg which we can pass in
+        streets = dict(
+            zip_longest(
+                ["street1", "street2", "street3"],
+                addr.street if addr is not None else [],
+                fillvalue=None,
+            )
+        )
+        desired_contact = PublicContact(
+            domain=self,
+            contact_type=contact_type,
+            registry_id=contact_id,
+            email=contact.email or "",
+            voice=contact.voice or "",
+            fax=contact.fax,
+            name=postal_info.name or "",
+            org=postal_info.org,
+            # For linter - default to "" instead of None
+            pw=getattr(auth_info, "pw", ""),
+            city=getattr(addr, "city", ""),
+            pc=getattr(addr, "pc", ""),
+            cc=getattr(addr, "cc", ""),
+            sp=getattr(addr, "sp", ""),
+            **streets,
+        )
+
+        return desired_contact
+
+    def _request_contact_info(self, contact: PublicContact):
```

Comment:
This is used in the `_registrant_to_public_contact()` function which exists on cache. The reason why we have this is because a registrant is the same type as other contacts. For registrant specifically, EppLib will return a string which is the id of the InfoContact object. For instance, if I had a contact object with id "eggs" then registrant="eggs". This isn't enough data for a PublicContact object, so we need to run the InfoContact command for more information to do the mapping. That coerced PublicContact object will then exist on the cache after it is 'converted' from its original form

Here is the code (clearer as a screenshot):
![image](https://github.com/cisagov/getgov/assets/141044360/1e74b151-605f-4082-9102-a6d68be01555)
![image](https://github.com/cisagov/getgov/assets/141044360/90a347d5-dd86-4afd-bd75-785c551fa015)

Its job is to effectively do the same thing that is being done here:
![image](https://github.com/cisagov/getgov/assets/141044360/ae832327-6f7c-4948-b2e2-bd93be09f2c5)

Reply:
Oh so the fetch contacts in cache is divided up in 2 blocks, one for registrant and one for the other 3?

Reply:
@rachidatecs Correct! Registrant is an edge case. It exists on a different property ("registrant" instead of "contacts") and EPPLib just returns a list of strings, instead of a list of DomainContact objects. I originally had this all contained inside one function but it just became harder to read, so I split it out